src/test/java/dmytro/hadiuchko/springboot/controller/BookControllerTest.java

```diff
@@ -0,0 +1,139 @@
+    @Mock
+    private BookService bookService;
+
+    @BeforeAll
+    static void beforeAll(@Autowired WebApplicationContext applicationContext) {
+        mockMvc = MockMvcBuilders
+                .webAppContextSetup(applicationContext)
+                .apply(springSecurity())
+                .build();
+    }
+
+    @Test
+    @WithMockUser(username = "admin", roles = {"ADMIN"})
+    @DisplayName("create a new book")
+    void createBook_ValidRequestDto_Success() throws Exception {
+        CreateBookRequestDto requestDto = new CreateBookRequestDto();
+        requestDto.setTitle(NEW_TITLE);
+        requestDto.setAuthor(NEW_AUTHOR);
+        requestDto.setPrice(NEW_PRICE);
+        requestDto.setIsbn(NEW_ISBN);
+        requestDto.setCoverImage(NEW_COVER_IMAGE);
+        requestDto.setDescription(NEW_DESCRIPTION);
+
+        BookDto expected = new BookDto();
+        expected.setId(BOOK_ID);
+        expected.setTitle(requestDto.getTitle());
+        expected.setAuthor(requestDto.getAuthor());
+        expected.setPrice(requestDto.getPrice());
+        expected.setIsbn(requestDto.getIsbn());
+        expected.setDescription(requestDto.getDescription());
+        expected.setCoverImage(requestDto.getCoverImage());
+
+        String jsonRequest = objectMapper.writeValueAsString(requestDto);
+
+        MvcResult result = mockMvc.perform(post(BOOK_URL)
+                        .content(jsonRequest)
+                        .contentType(MediaType.APPLICATION_JSON)
+                )
+                .andExpect(status().isCreated())
+                .andReturn();
+        BookDto actual = objectMapper.readValue(result.getResponse()
+                .getContentAsString(), BookDto.class);
+
+        assertNotNull(actual);
+        EqualsBuilder.reflectionEquals(expected, actual, "id");
+    }
+
+    @Test
+    @WithMockUser(username = "user", roles = {"ADMIN", "USER"})
+    @DisplayName("Return all books with valid role")
+    void getAllBooks_withUserRole_ReturnAll() throws Exception {
+        List<BookDto> bookDtoList = List.of(new BookDto());
+        when(bookService.findAll(any(Pageable.class))).thenReturn(bookDtoList);
+
+        mockMvc.perform(get(BOOK_URL))
+                .andExpect(status().isOk());
+    }
+
+    @Test
+    @WithMockUser(username = "user", roles = {"ADMIN", "USER"})
+    @DisplayName("Return book by id with valid role")
+    void getBookById_withUserRole_ReturnBook() throws Exception {
+        BookDto bookDto = new BookDto();
+        bookDto.setId(BOOK_ID);
+        bookDto.setTitle(BOOK_TITLE);
+
+        when(bookService.findById(BOOK_ID)).thenReturn(bookDto);
+        mockMvc.perform(get(BOOK_BY_ID_URL))
+                .andExpect(status().isOk());
+    }
+
+    @Test
+    @WithMockUser(username = "admin", roles = "ADMIN")
+    @DisplayName("Delete book by id with valid role")
+    void deleteBook_WithAdminRole_Success() throws Exception {
+        doNothing().when(bookService).deleteById(BOOK_ID);
```

Comment:
again let's not mock any behavior in integrational tests, and try to make them as real as it gets